test/NetTopologySuite.Samples.Console/Tests/Github/Issue447.cs

```diff
@@ -13,18 +14,78 @@ public class Issue447
+            Assert.IsFalse(pt1 == pt3);
+            Assert.IsFalse(pt1 == pt4);
+            Assert.IsFalse(pt1 == pt5);
+            Assert.IsTrue(pt1 == pt6);
             Assert.AreNotEqual(pt1, pt2);
+            Assert.AreNotEqual(pt1, pt3);
+            Assert.AreNotEqual(pt1, pt4);
+            Assert.AreNotEqual(pt2, pt3);
+            Assert.AreNotEqual(pt2, pt4);
+            Assert.AreNotEqual(pt3, pt4);
+            Assert.AreNotEqual(pt4, pt5);
+            Assert.AreEqual(pt1, pt6);
+
+            Assert.IsTrue(pt1 == f.CreatePoint(new CoordinateZ(1,2)));
+            Assert.IsTrue(pt1 == f.CreatePoint(new CoordinateM(1, 2)));
+            Assert.IsTrue(pt1 == f.CreatePoint(new CoordinateZM(1, 2)));
+
+        }
+
+        [Test]
+        public void TestWithTolerance()
+        {
+            var i = new NtsGeometryServices(CsFactory, PrecisionModel.Floating.Value, 0,
+                GeometryOverlay.Legacy, new PerOrdinateEqualityComparer());
+            var f = i.CreateGeometryFactory();
+            const double tolerance  = 0.001;
+            const double allowed = tolerance;
+
+            var pt = f.CreatePoint(new CoordinateZM(1, 2, 3, 4));
+            Assert.That(pt == f.CreatePoint(new CoordinateZM(1 + allowed, 2, 3, 4)), Is.False);
+            Assert.That(pt.EqualsExact(f.CreatePoint(new CoordinateZM(1 + allowed, 2, 3, 4)), tolerance), Is.True);
+            Assert.That(pt.EqualsExact(f.CreatePoint(new CoordinateZM(Increment(1 + allowed), 2, 3, 4)), tolerance), Is.False);
+
+            Assert.That(pt == f.CreatePoint(new CoordinateZM(1 + allowed, 2, 3, 4)), Is.False);
+            Assert.That(pt.EqualsExact(f.CreatePoint(new CoordinateZM(1, 2 + allowed, 3, 4)), tolerance), Is.True);
+            Assert.That(pt.EqualsExact(f.CreatePoint(new CoordinateZM(1, Increment(2 + allowed), 3, 4)), tolerance), Is.False);
+
+            Assert.That(pt == f.CreatePoint(new CoordinateZM(1 + allowed, 2, 3, 4)), Is.False);
+            Assert.That(pt.EqualsExact(f.CreatePoint(new CoordinateZM(1, 2, 3 - allowed, 4)), tolerance), Is.True);
+            Assert.That(pt.EqualsExact(f.CreatePoint(new CoordinateZM(1, 2, Decrement(3 - allowed), 4)), tolerance), Is.False);
+
+            Assert.That(pt == f.CreatePoint(new CoordinateZM(1 + allowed, 2, 3, 4)), Is.False);
+            Assert.That(pt.EqualsExact(f.CreatePoint(new CoordinateZM(1, 2, 3, 4 - allowed)), tolerance), Is.True);
+            Assert.That(pt.EqualsExact(f.CreatePoint(new CoordinateZM(1, 2, 3, Decrement(4 - allowed))), tolerance), Is.False);
+        }
+
+        private static double Increment(double value)
+        {
+            return Math.Sign(value) * BitConverter.Int64BitsToDouble(BitConverter.DoubleToInt64Bits(Math.Abs(value)) + 1);
+        }
+        private static double Decrement(double value)
+        {
+            return Math.Sign(value) * BitConverter.Int64BitsToDouble(BitConverter.DoubleToInt64Bits(Math.Abs(value)) - 1);
         }
```

Comment:
I think multiplying by `Math.Sign` might be wrong for negative values (which we don't have, so we wouldn't know)... regardless, we can use [Math.BitIncrement](https://docs.microsoft.com/en-us/dotnet/api/system.math.bitincrement) and [Math.BitDecrement](https://docs.microsoft.com/en-us/dotnet/api/system.math.bitdecrement) in our tests these days.

Reply:
👍 . I'm always amazed how well you know the dotnet api and its libraries.
Thanks